lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -316,16 +317,30 @@ def _find_matches(self, requirements):
         # The fqcn is guaranteed to be the same
         version_req = "A SemVer-compliant version or '*' is required. See https://semver.org to learn how to compose it correctly. "
         version_req += "This is an issue with the collection."
-        try:
-            coll_versions = self._api_proxy.get_collection_versions(first_req)
-        except TypeError as exc:
-            if first_req.is_concrete_artifact:
-                # Non hashable versions will cause a TypeError
-                raise ValueError(
-                    f"Invalid version found for the collection '{first_req}'. {version_req}"
-                ) from exc
-            # Unexpected error from a Galaxy server
-            raise
+
+        # If we're upgrading collections, we can't calculate preinstalled_candidates until the latest matches are found.
+        # Otherwise, we can potentially avoid a Galaxy API call by doing this first.
+        preinstalled_candidates = None
+        if not self._upgrade and first_req.type == 'galaxy':
+            preinstalled_candidates = {
+                candidate for candidate in self._preferred_candidates
+                if candidate.fqcn == fqcn and
+                all(self.is_satisfied_by(requirement, candidate) for requirement in requirements)
+            }
+        if preinstalled_candidates:
+            # Since we're not upgrading and a sufficient match was found, don't look for more recent versions
+            coll_versions = []  # type: t.Iterable[t.Tuple[str, GalaxyAPI]]
+        else:
+            try:
+                coll_versions = self._api_proxy.get_collection_versions(first_req)
+            except TypeError as exc:
+                if first_req.is_concrete_artifact:
+                    # Non hashable versions will cause a TypeError
+                    raise ValueError(
+                        f"Invalid version found for the collection '{first_req}'. {version_req}"
+                    ) from exc
+                # Unexpected error from a Galaxy server
+                raise
```

Comment:
How about keeping the nesting low?
```suggestion
        try:
            coll_versions = [] if preinstalled_candidates else self._api_proxy.get_collection_versions(first_req)  # type: t.Iterable[t.Tuple[str, GalaxyAPI]]
        except TypeError as exc:
            if first_req.is_concrete_artifact:
                # Non hashable versions will cause a TypeError
                raise ValueError(
                    f"Invalid version found for the collection '{first_req}'. {version_req}"
                ) from exc
            # Unexpected error from a Galaxy server
            raise
```